AI/MMAI/BAI/v13/stack.cpp

```diff
@@ -0,0 +1,543 @@
+/*
+ * stack.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "StdInc.h"
+#include "CCreatureHandler.h"
+#include "GameLibrary.h"
+#include "battle/IBattleInfoCallback.h"
+#include "bonuses/BonusEnum.h"
+#include "constants/EntityIdentifiers.h"
+
+#include "BAI/v13/stack.h"
+#include "schema/v13/constants.h"
+#include "schema/v13/types.h"
+
+namespace MMAI::BAI::V13
+{
+namespace S13 = Schema::V13;
+using SA = Schema::V13::StackAttribute;
+using F1 = Schema::V13::StackFlag1;
+using F2 = Schema::V13::StackFlag2;
+using GA = Schema::V13::GlobalAttribute;
+
+namespace
+{
+	int calculateSlot(const CStack * cstack)
+	{
+		int slot = cstack->unitSlot();
+		if(slot >= 0 && slot < 7)
+		{
+			return slot;
+		}
+		else if(slot == SlotID::WAR_MACHINES_SLOT)
+		{
+			return S13::STACK_SLOT_WARMACHINES;
+		}
+		else
+		{
+			// "special" slot, e.g. summoned, commander, etc.
+			return S13::STACK_SLOT_SPECIAL;
+		}
+	}
+
+	char calculateAlias(int slot)
+	{
+		switch(slot)
+		{
+			case S13::STACK_SLOT_SPECIAL:
+				return 'S';
+				break;
+			case S13::STACK_SLOT_WARMACHINES:
+				return 'M';
+				break;
+			default:
+				return static_cast<char>('0' + slot);
+		}
+	}
+}
+
+// static
+int Stack::CalcValue(const CCreature * cr)
+{
+	static const bool verbose = []()
+	{
+		const char * envvar = std::getenv("MMAI_VERBOSE");
+		return envvar != nullptr && std::strcmp(envvar, "1") == 0;
+	}();
+
+	static std::map<const CCreature *, float> cache;
```

Comment:
Any way to avoid non-const static variable? 

Perhaps compute score for each in-game unit on AI initialization and store it as part of AI itself?
This would ensure that value is guaranteed to exists and will also allow dropping mutex guards

Reply:
If there's a way to precompute score for each unit once and store it in a const static var, that would be great.
Can you help me with an approach to iterate through in-game units? I was unable figure it out :(

Reply:
See here for working with creatures: https://github.com/vcmi/vcmi/blob/b6c19f4895392e0440ead08386a34c04dea43ae1/client/lobby/BattleOnlyModeTab.cpp#L561

If you want all (not only DefaultAllowed): Iterate over `LIBRARY->creh->objects`.

Does this help?

Reply:
I will try this, thanks

Reply:
Done 👍